src/hooks/useForm.js

```diff
@@ -0,0 +1,218 @@
+import { useState } from 'react';
+
+import { validate } from '../validation';
+import { EMPTY_VALUE } from '../validation/constants';
+
+/**
+ * The useForm hook return value.
+ * @typedef {object} UseFormReturnValue
+ * @property {Object.<string,string>} form - The values of the properties in the form.
+ * @property {Object.<string,string[]>} errors - The errors in the form.
+ * @property {string} submitError - The error message when the form fails to submit successfully.
+ * @property {Function} handleReset - Function to reset the form to its initial state.
+ * @property {Function} handleSubmit - Function to validate the form and call the user callback.
+ * @property {Function} handleInputChange - Function to validate changed input and set the state.
+ */
+
+/**
+ * The useForm state.
+ * @typedef {object} UseFormState
+ * @property {Object.<string,string>} form - The values of the properties in the form.
+ * @property {Object.<string,string[]>} errors - The errors in the form.
+ * @property {string} submitError - The error message when the form fails to submit successfully.
+ */
+
+/**
+ * Handle form functionality.
+ * @class
+ * @param {Function} schema - The schema of the form.
+ * @param {object} [initialFormState=null] - The default values for form elements.
+ * @returns {UseFormReturnValue} The form, errors and handlers.
+ *
+ * @example
+ * const {
+ *  form,
+ *  errors,
+ *  submitError,
+ *  handleReset,
+ *  handleSubmit,
+ *  handleInputChange
+ * } = useForm(formSchema);
+ */
+function useForm(schema, initialFormState = null) {
+  /************************************
+   * State
+   ************************************/
+
+  const [state, setState] = useState(init(schema, initialFormState));
```

Comment:
it seems weird having a state named state in a functional component but i see no reason not to

Reply:
Yup, that's my throwback to class components. Although usually it's best to have individual pieces, combining the state made it much easier to work with.